File: pt5/1143.longest-common-subsequence.py
from __future__ import annotations


class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        if len(text2) > len(text1):
            text1, text2 = text2, text1

        dp = [0] * (len(text2) + 1)
        for char1 in text1:
            previous_diagonal = 0
            for j, char2 in enumerate(text2, start=1):
                saved = dp[j]
                if char1 == char2:
                    dp[j] = previous_diagonal + 1
                else:
                    dp[j] = max(dp[j], dp[j - 1])
                previous_diagonal = saved

        return dp[-1]

File: pt5/1143.longest-common-subsequence.go
package main

func longestCommonSubsequence(text1 string, text2 string) int {
	if len(text2) > len(text1) {
		text1, text2 = text2, text1
	}

	dp := make([]int, len(text2)+1)
	for i := 0; i < len(text1); i++ {
		previousDiagonal := 0
		for j := 1; j <= len(text2); j++ {
			saved := dp[j]
			if text1[i] == text2[j-1] {
				dp[j] = previousDiagonal + 1
			} else if dp[j-1] > dp[j] {
				dp[j] = dp[j-1]
			}
			previousDiagonal = saved
		}
	}

	return dp[len(text2)]
}

